edu_edfi_airflow/callables/change_version.py

```diff
@@ -0,0 +1,290 @@
+import logging
+
+from typing import Dict, List, Tuple, Optional
+
+from airflow.exceptions import AirflowSkipException, AirflowFailException
+from airflow.providers.snowflake.hooks.snowflake import SnowflakeHook
+
+from edu_edfi_airflow.callables.snowflake import insert_into_snowflake
+from edu_edfi_airflow.callables import airflow_util
+from edu_edfi_airflow.providers.edfi.hooks.edfi import EdFiHook
+
+
+def get_newest_edfi_change_version(edfi_conn_id: str, **kwargs) -> int:
+    """
+
+    :return:
+    """
+    ### Connect to EdFi ODS and verify EdFi3.
+    edfi_conn = EdFiHook(edfi_conn_id=edfi_conn_id).get_conn()
+
+    # Break off prematurely if change versions not supported.
+    if edfi_conn.is_edfi2():
+        logging.warning("Change versions are only supported in EdFi 3+!")
+        return None
+
+    # Pull current max change version from EdFi.
+    max_change_version = edfi_conn.get_newest_change_version()
+    logging.info(f"Current max change version is `{max_change_version}`.")
+
+    return max_change_version
+
+
+def reset_change_versions(
+    tenant_code : str,
+    api_year    : int,
+
+    *,
+    snowflake_conn_id: str,
+    change_version_table: str,
+
+    **kwargs
+) -> None:
+    # Airflow-skip if run not marked for a full-refresh.
+    if not airflow_util.is_full_refresh(kwargs):
+        raise AirflowSkipException(f"Full refresh not specified. Change version table `{change_version_table}` unchanged.")
+
+    ### Prepare the SQL query.
+    # Retrieve the database and schema from the Snowflake hook, and raise an exception if undefined.
+    database, schema = airflow_util.get_snowflake_params_from_conn(snowflake_conn_id)
+
+    # Reset all resources in this tenant-year.
+    qry_mark_inactive = f"""
+        update {database}.{schema}.{change_version_table}
+            set is_active = FALSE
+        where tenant_code = '{tenant_code}'
+        and api_year = {api_year}
+        and is_active
+    """
+
+    # Filter only to inactive endpoints to those specified in DAG-configs, if defined.
+    if config_endpoints := airflow_util.get_config_endpoints(kwargs):
+        qry_mark_inactive += "    and name in ('{}')".format("', '".join(config_endpoints))
+
+    ### Connect to Snowflake and execute the query.
+    logging.info("Full refresh: marking previous pulls inactive.")
+    SnowflakeHook(snowflake_conn_id).run(qry_mark_inactive)
+
+
+def get_previous_change_versions(
```

Comment:
Function `get_previous_change_versions()` has three main differences:
* Skip if a deletes or keyChanges run in the context of a full-refresh.
* Add keyChanges processing.
* Push xcoms as a single object instead of individual xcoms.